do not squash:  docs(furo theme):  implement

vwheeler63 opened the pull request:
Generated docs in Furo theme [can be found here](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/furo/).

- updated `conf.py` to go with Furo theme;
- removed hidden `.. toctree::` directive in stop `index.rst`—not compatible with Furo theme;
- FontAwesome included manually and set up so open/close icons work correctly (e.g. in API pages, and source code in examples) (This font was previously added automatically by RTD theme.);
- added some `.css` files relevant to FontAwesome (these were previously added automatically by RTD theme);
- manually added `jQuery` library (previously added automatically by RTD theme—needed for some custom JavaScript to run correctly) (note:  we can't use `sphinxcontrib-jquery` because it is adding jQuery.js at end of HTML instead of beginning—seems to be related to the theme switch);
- docs-version drop-down list now works with Furo theme;
- styled docs-version drop-down list to look nicer (width, position);
- styled docs-version drop-down list to comply with the `color` and `background-color` styles used by the theme (so now it switches with the theme);
- removed remnants of RTD theme in our `custom.css`;
- removed `layout.html` template, which does not apply to Furo theme;
- cleaned up the JavaScript code to make it more understandable (eliminated re-use of the name "versions" for several different things, causing confusion);
- documented custom JavaScript in the `page.html` template (in a way that the documentation is only in the template, and not in the production HTML pages); and
- re-generated the [Furo example docs](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/furo/) so reviewers could have a look at them.

cc:  @kisvegabor 

Note:  I believe I can easily add the [Prev] and [Next] buttons at the top of the pages as well.  Would you like me to?

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.  Done.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.  n/a
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.  n/a
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).  n/a
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).  n/a
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.  Done.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.  Done.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

liamHowatt reviewed (dismissed):
Awesome!

The light/dark theme toggle has a 3rd mode which seems to be the same as light.

It seems like the version dropdown has a lot of code that won't be tested until it's merged. Or was it tested somehow?

vwheeler63 commented:
> The light/dark theme toggle has a 3rd mode which seems to be the same as light.

That third mode mode is called "auto" and internally it matches whatever the OS is set for.  I cannot speak for Linux, but as of Windows 10, the user can tell the OS he prefers "light" or "dark" theme.  After that, all the "compliant" applications and websites follow with their coloring schemes.  In the Furo theme it is done via a CSS Media Query (done through JavaScript instead), which internally queries the OS for the setting, and returns a Boolean based on the query string.  It looks like this:
```javascript
  const currentTheme = localStorage.getItem("theme") || "auto";
  const prefersDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
```
With those 2 values, the JavaScript then toggles the theme accordingly.  On my Windows system, I have it set up to "prefer dark", and so that 3rd mode is dark on my system.

Unfortunately, the algorithm in the Furo theme does not assume that the user might change the OS setting in the middle of a browsing session, or perhaps even that the OS will change it (perhaps at a certain time of day), so the query is ONLY done when the button is pressed,  (I might have done it when each page is loaded/refreshed, for when OSes get fancier and allow the user to specify WHEN he wants "dark mode".)
 
> It seems like the version drop-down has a lot of code that won't be tested until it's merged. Or was it tested somehow?

I tested it in 2 ways.  It's the same code that was used previously, only it stopped appearing in Furo.  I had to find out why and remedy (it was assuming an HTML-element structure and classes that were part of the `sphinx_rtd_theme`), so I merely made adjustments to work with the Furo HTML-element structure and classes.  All else is unchanged.

The delay getting this PR submitted was due to the fact that I decided to solve it myself by getting myself trained up through "solid basics" in CSS and client-side JavaScript with an bit of HTML3 -> HTML5 update.

The tests performed are:

- locally watching the JavaScript execute:  when it fires, it triggers assembly of a URL and tells the window to navigate to it:
  ```
  protocol + host + /version_no/
  //e.g.
  https://docs.lvgl.io/9.2/
  ```
  which still works.  It's not a very portable algorithm (assumes a certain directory structure), but I saw no reason to change it (for now).

- The 2nd test was to build a directory structure on my host web-space that matched the algorithm, and verified it worked there, both:
    - loading (when the web pages is loaded, it populates the drop-down list's "value", and
    - after selecting between "master" and "9.2" that it switches URLs accordingly.

Note1:  I switched the initial hosting web location to http://crystal-clear-research.com/master/ to comply with the version-switch algorithm.  However, I only set up "master" and "9.2" directories, so "9.1" and below won't work for this demo, but will work on https://docs.lvgl.io where all of those versions are hosted.

Note2:  I found one more thing I need to handle so I just switched this PR back to "draft" mode.

vwheeler63 commented:
Note:  I may need some help here....  (I'm really glad we got the doc-build check in here!)

While the new 'furo' theme tests have been building fine on my system, the current doc-build on Ubuntu is complaining with this message at the base of the traceback:

    sphinx.errors.ThemeError: no theme named 'furo' found (missing theme.toml?)

I'm puzzled because the 'furo' theme has COME WITH Sphinx at least since I became aware of it and I did tests with it as far back as Sept 2024 (and probably much longer).  And Sphinx was version 7.x back then -- now it is v8.x and the same appears true with the installed version on this test in Ubuntu:  when it runs, Sphinx announces its version:  "v8.2.3".  Is it possible that the 'furo' theme is not coming with Sphinx on Ubuntu?  @kisvegabor , did you have to do anything special on your system to get 'furo' working?

@liamHowatt , any thoughts?

vwheeler63 commented:
Note:  I "called for help" earlier, but realized what the problem was and fixed it, so I deleted the message -- in case you are arriving here to a comment URL that is no longer here.  :-)

vwheeler63 commented:
@kisvegabor If you haven't already done so, Gábor, I need the final URL for the dynamic banner JSON file -- then this PR will be ready to merge.

kisvegabor commented:
@vwheeler63 does the URL we shared in email work?

vwheeler63 commented:
> @vwheeler63 does the URL we shared in email work?

Yes!  What I did to test it (given the CORS restriction on the `fetch()` call) is a copied exactly that `.json` file to http://crystal-clear-research.com/master/banner.json  and tested it where you can see the results here:  http://crystal-clear-research.com/master/

Then I switched the URL to the one Richard gave me and submitted commit 2d3e6a0 with that URL and confirmed it was typed correctly by simply directing my browser to that `.json` file.

So it will work once the files are being hosted on the `docs.lvgl.io` domain (the `fetch()` will be fetching from the same domain, thus not violating the CORS restriction).

vwheeler63 commented:
> @vwheeler63 does the URL we shared in email work?

Additionally, I just saw that the `banner.json` file's content changed to `[ { } ]`.  I previously coded for an empty array, but not an array with an empty object(s), so I made some adjustments in `custom.js` to handle both gracefully, and those adjustments are now in this PR.

vwheeler63 commented:
> So it just downloads `banner.json` when the site is loaded and reads it, right? My concerns is: will it be updated in the docs without rebuilding the docs if we change `banner.json`?

Exactly!  `custom.js` hooks the event that fires when the full HTML page is loaded, and when that event fires, it does a `fetch()` (async server pull) from the URL Richard provided, parses it into a JSON object and acts according to its content.  None of that gets stored anywhere in the generated docs.

kisvegabor commented:
Fantastic! Thanks you! We should work on the style of the banner a little before mering, but it's not a big deal. Just a good color, and minor other CSS things. 

@richardgazdik later this week please send a CSS class to @vwheeler63 for the banner [here](http://crystal-clear-research.com/master/).

vwheeler63 commented:
@kisvegabor Update:  Good news....  The (sabotaged) Field Lists format appears to be controllable inside the style sheet....  Looking deeper now.

vwheeler63 commented:
@kisvegabor Now I'm happy again!  The last commit does it:

![image](https://github.com/user-attachments/assets/6c84efde-6fd3-455e-9bd3-16a1d5fc7f17)

rather than this:

![image](https://github.com/user-attachments/assets/1b5e7d7b-ecee-47de-864f-14b8a3ac0f51)

Okay -- my tail is wagging about Furo again.  😄

vwheeler63 commented:
Rebased in order to resolve conflict in `custom.css`.

kisvegabor commented:
Sorry for my absence. In the meantime I've discussed with @richardgazdik the probably he can improve the styles and fix a few visual things in a few days. 


---

It's great that now we cleary see how complex some docs are:
![image](https://github.com/user-attachments/assets/31cfe4a8-35db-44ab-b71a-d8430d0fcdf9)

E.g. Observers should be 3-4 separate pages. But it's for  a new round.

vwheeler63 commented:
Hi, @kisvegabor !

> ... he can improve the styles and fix a few visual things in a few days.

Sweet!  What will be the mechanics?  Merge this one and he starts his own branch, or would he like to clone my fork for push to this branch?  (I am open for either.)

kisvegabor commented:
Maybe @richardgazdik can push his changes to his fork and you can cherry-pick his commit to this PR. What do you think?

vwheeler63 commented:
> Maybe @richardgazdik can push his changes to his fork and you can cherry-pick his commit to this PR. What do you think?

I'm not sure how to make GIT cherry pick between repositories, but I could certainly look at the diffs and implement his changes that way.  

What "base" is he pushing changes to?  The state of this branch?

kisvegabor commented:
The cherry-picking magic should be something like this on your end:
1. `git checkout docs/switch_to_furo_theme_02`
1. `git checkout -b rich-design` to create a new branch from your code
2. `git pull https://github.com/richardgazdik/lvgl_fork design_fixes --rebase` get the updates from Rich's fork
1. `git log -5` to see the commit hashes
1. `git checkout docs/switch_to_furo_theme_02` to go back to your branch
1. `git cherry-pick some_hash`

Maybe there is a simpler way, but that's how I would/could do it.

liamHowatt commented:
I would do it like

1. `git checkout docs/switch_to_furo_theme_02`
2. `git remote add richardgazdik https://github.com/richardgazdik/lvgl_fork`
3. `git fetch richardgazdik`
4. Visit https://github.com/richardgazdik/lvgl_upstream/commits/design_fixes/ in my browser to view the commits+hashes
5. `git cherry-pick some_hash`

vwheeler63 commented:
@liamHowatt @kisvegabor Brilliant!  `**feeling silly**``  I should have thought of that....  I did something similar with Kevin's changes he was working back in Aug-Sep-2024.

Is Richard's work ready for me to do that now?

kisvegabor commented:
Update! 2 hours remained :slightly_smiling_face:

vwheeler63 commented:
> Update! 2 hours remained 🙂

cc:  Richard (but I can't find his name)

Okay!  If I am understanding that correctly, it is time for me to act?  If so, what is the name of the branch I need to use?

I see a `https://github.com/richardgazdik/lvgl_upstream/tree/docs/switch_to_furo_theme_03` that looks like it might be the one.  Can you confirm?

liamHowatt commented:
I just spoke to him a moment ago. He found more things he wants to change. It's not ready yet. He wants to finish it tonight. He said he may comment here. Yes you found his branch :smile: nice.

vwheeler63 commented:
> I just spoke to him a moment ago. He found more things he wants to change. It's not ready yet. He wants to finish it tonight. He said he may comment here. Yes you found his branch 😄 nice.

Cool!  Yes, I brought it into my local repository and saw that it was EXTENDED from my `docs/switch_to_furo_theme_02` branch, and then it was obvious!  :-)

Meanwhile I will be looking at the changes he has made so far.  If Richard doesn't see this message himself, just tell him to be sure to extend THAT branch and we will be in perfect coordination!  😄    (Goodness, I love Git!!!)

kisvegabor commented:
We have the feature freeze week this week. If all goes well, we have the release early next week, so we can merge this soon!

kisvegabor commented:
Sounds interesting! Let's target next week, we are swamped with work now. :face_with_spiral_eyes:

vwheeler63 commented:
> Sounds interesting! Let's target next week, we are swamped with work now. 😵‍💫

Okay!
